syntax = "proto3";

package vector;

service VectorService {
  rpc EmbedText (EmbedRequest) returns (EmbedResponse);
  rpc Search (SearchRequest) returns (SearchResponse);
  rpc Upsert (UpsertRequest) returns (UpsertResponse);
}

message UpsertRequest {
  string id = 1;
  string text = 2;
  map<string, string> metadata = 3;
}

message UpsertResponse {
  bool success = 1;
  string error = 2;
}

message EmbedRequest {
  string text = 1;
}

message EmbedResponse {
  repeated float vector = 1;
}

message SearchRequest {
  string query_text = 1;
  int32 top_k = 2;
  float min_score = 3;
}

message SearchResult {
  string id = 1;
  float score = 2;
  string content = 3;
  map<string, string> metadata = 4;
}

message SearchResponse {
  repeated SearchResult results = 1;
}
